### PR TITLE
Inttest adding tests using big fileids

### PR DESCRIPTION
**Pull request by SergioBertolinSG**
Altering most of relevant to file_id columns to be BIGINT.
Ref https://github.com/owncloud/core/issues/26901

With integration tests.

Not sure if I am in the right path, @PVince81 what do you think?

### CODE REVIEW

**Comment by SergioBertolinSG** [comment]
After using this app and doing 

curl -X POST http://admin:admin@10.40.40.182/html/multilinkshares/ocs/v1.php/apps/testing/api/v1/increasefileid


Checking oc_filecache table in the database I see the entry correctly added. But when accessing the server there is an exception:

```
"reqId":"G38RbN69ys8hW1Kqptl7","level":3,"time":"2017-07-11T10:58:55+00:00","remoteAddr":"10.40.40.134","user":"admin","app":"PHP","method":"POST","url":"\/html\/multilinkshares\/ocs\/v1.php\/apps\/testing\/api\/v1\/increasefileid","message":"Error: Call to a member function succeeded() on null at \/var\/www\/html\/multilinkshares\/lib\/private\/legacy\/api.php#214"}
{"reqId":"xinaIyhR1BW9s0u1R4CJ","level":3,"time":"2017-07-11T10:59:27+00:00","remoteAddr":"10.40.40.134","user":"admin","app":"index","method":"GET","url":"\/html\/multilinkshares\/index.php\/apps\/files\/","message":"Exception: {\"Exception\":\"Doctrine\\\\DBAL\\\\Exception\\\\UniqueConstraintViolationException\",\"Message\":\"An exception occurred while executing 'INSERT INTO `oc_filecache` (`mimepart`,`mimetype`,`mtime`,`size`,`etag`,`storage_mtime`,`permissions`,`checksum`,`path_hash`,`path`,`parent`,`name`,`storage`) SELECT ?,?,?,?,?,?,?,?,?,?,?,?,? FROM `oc_filecache` WHERE `storage` = ? AND `path_hash` = ? HAVING COUNT(*) = 0' with params [\\\"1\\\", \\\"2\\\", 1499770764, -1, \\\"5964af8f94471\\\", 1499770764, 23, \\\"\\\", \\\"d41d8cd98f00b204e9800998ecf8427e\\\", \\\"\\\", -1, \\\"\\\", 2, 2, \\\"d41d8cd98f00b204e9800998ecf8427e\\\"]:\\n\\nSQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '2147483647' for key 'PRIMARY'\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/composer\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/DBALException.php(128): Doctrine\\\\DBAL\\\\Driver\\\\AbstractMySQLDriver->convertException('An exception oc...', Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException))\\n#1 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/composer\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Connection.php(1015): Doctrine\\\\DBAL\\\\DBALException::driverExceptionDuringQuery(Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOMySql\\\\Driver), Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException), 'INSERT INTO `oc...', Array)\\n#2 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/DB\\\/Connection.php(211): Doctrine\\\\DBAL\\\\Connection->executeUpdate('INSERT INTO `oc...', Array, Array)\\n#3 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/DB\\\/Adapter.php(113): OC\\\\DB\\\\Connection->executeUpdate('INSERT INTO `oc...', Array)\\n#4 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/DB\\\/Connection.php(249): OC\\\\DB\\\\Adapter->insertIfNotExist('*PREFIX*filecac...', Array, Array)\\n#5 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/Files\\\/Cache\\\/Cache.php(265): OC\\\\DB\\\\Connection->insertIfNotExist('*PREFIX*filecac...', Array, Array)\\n#6 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/Files\\\/Cache\\\/Cache.php(222): OC\\\\Files\\\\Cache\\\\Cache->insert('', Array)\\n#7 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/Files\\\/Cache\\\/Scanner.php(264): OC\\\\Files\\\\Cache\\\\Cache->put('', Array)\\n#8 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/Files\\\/Cache\\\/Scanner.php(204): OC\\\\Files\\\\Cache\\\\Scanner->addToCache('', Array, -1)\\n#9 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/Files\\\/Cache\\\/Scanner.php(168): OC\\\\Files\\\\Cache\\\\Scanner->scanFile('')\\n#10 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/Files\\\/Cache\\\/Scanner.php(307): OC\\\\Files\\\\Cache\\\\Scanner->scanFile('avatars', 3, -1, NULL, true)\\n#11 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/Files\\\/View.php(1302): OC\\\\Files\\\\Cache\\\\Scanner->scan('avatars', false)\\n#12 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/Files\\\/View.php(1343): OC\\\\Files\\\\View->getCacheEntry(Object(OCA\\\\Files_Trashbin\\\\Storage), 'avatars', '\\\/avatars')\\n#13 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/Files\\\/Node\\\/Root.php(182): OC\\\\Files\\\\View->getFileInfo('\\\/avatars')\\n#14 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/AvatarManager.php(97): OC\\\\Files\\\\Node\\\\Root->get('\\\/avatars')\\n#15 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/AvatarManager.php(117): OC\\\\AvatarManager->getFolder(Object(OC\\\\Files\\\\Node\\\\Root), 'avatars')\\n#16 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/AvatarManager.php(91): OC\\\\AvatarManager->getAvatarFolder(Object(OC\\\\User\\\\User))\\n#17 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/TemplateLayout.php(101): OC\\\\AvatarManager->getAvatar('admin')\\n#18 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/legacy\\\/template.php(232): OC\\\\TemplateLayout->__construct('user', 'files')\\n#19 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/public\\\/AppFramework\\\/Http\\\/TemplateResponse.php(156): OC_Template->fetchPage()\\n#20 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/AppFramework\\\/Http\\\/Dispatcher.php(112): OCP\\\\AppFramework\\\\Http\\\\TemplateResponse->render()\\n#21 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/AppFramework\\\/App.php(98): OC\\\\AppFramework\\\\Http\\\\Dispatcher->dispatch(Object(OCA\\\\Files\\\\Controller\\\\ViewController), 'index')\\n#22 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/AppFramework\\\/Routing\\\/RouteActionHandler.php(46): OC\\\\AppFramework\\\\App::main('ViewController', 'index', Object(OC\\\\AppFramework\\\\DependencyInjection\\\\DIContainer), Array)\\n#23 [internal function]: OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler->__invoke(Array)\\n#24 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/private\\\/Route\\\/Router.php(299): call_user_func(Object(OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler), Array)\\n#25 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/base.php(920): OC\\\\Route\\\\Router->match('\\\/apps\\\/files\\\/')\\n#26 \\\/var\\\/www\\\/html\\\/multilinkshares\\\/index.php(49): OC::handleRequest()\\n#27 {main}\",\"File\":\"\\\/var\\\/www\\\/html\\\/multilinkshares\\\/lib\\\/composer\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Driver\\\/AbstractMySQLDriver.php\",\"Line\":66}"}
```

**Comment by SergioBertolinSG** [comment]
OK, it is ready.

It increases the fileid in the before suite statement only if it is not present already.

@VicDeo This PR should wait for your fix. You could do it here or I'll rebase it after it is merged, as you wish.

**Comment by VicDeo** [comment]
a gift from pgsql https://stackoverflow.com/questions/43467722/postgresql-column-type-conversion-form-bigint-to-bigserial-with-liquibase-change

**Comment by VicDeo** [comment]
So...
`type "bigserial" does not exist` error on postgres. Because bigserial is not a real type. 
I'm stuck :(

**Comment by PVince81** [comment]
@VicDeo are you saying that postgres has no support for bigint ??

**Comment by VicDeo** [comment]
@PVince81 no, I say that design of doctrine/dbal leaves no room for a clean solution.
1. `bigserial` is a metatype. it can't be used in `ALTER blabla`  queries. 
2. `getBigIntTypeDeclarationSQL` is not aware of a query context. It always returns `bigserial` for bigints with autoincrement
3. 1+2 = schemaDiff produces broken SQL for Postrge

To prevent this I can skip this mogration for posgre and write plain SQL migration with a direct `ALTER ` query

**Comment by PVince81** [comment]
@DeepDiver1975 I'd like your input on this https://github.com/owncloud/core/pull/28364#issuecomment-315006853

We could also try to submit a PR upstream to Doctrine/dbal...

**Comment by tomneedham** [comment]
See https://github.com/owncloud/core/issues/28429

**Comment by VicDeo** [comment]
hmm. there is no crash locally

**Comment by VicDeo** [comment]
@PVince81 @DeepDiver1975 All tests are green.

File: apps/files_sharing/appinfo/database.xml [hunk 1]
```diff
@@ -13,7 +13,7 @@
 				<default>0</default>
 				<notnull>true</notnull>
 				<autoincrement>1</autoincrement>
-				<length>4</length>
+				<length>10</length>
```
Comment (PVince81) [inline apps/files_sharing/appinfo/database.xml]: hmm, I thought @phisch got rid of these files when he converted 8.2 stuff to migrations. But maybe these were older schemas that didn't need migrating

File: lib/private/DB/PostgreSqlMigrator.php [hunk 2]
```diff
@@ -52,4 +52,32 @@ protected function getDiff(Schema $targetSchema, \Doctrine\DBAL\Connection $conn
...
+
+		$connection->beginTransaction();
+		$sqls = $schemaDiff->toSql($connection->getDatabasePlatform());
+		$step = 0;
+		foreach ($sqls as $sql) {
+			$this->emit($sql, $step++, count($sqls));
+			// BIGSERIAL could not be used in statements altering column type
+			// That's why we replace it with BIGINT 
+			// see https://github.com/owncloud/core/pull/28364#issuecomment-315006853
+			if (preg_match('|(ALTER TABLE\s+\S+\s+ALTER\s+\S+\s+TYPE\s+)(BIGSERIAL)|i', $sql, $matches)){
```
Comment (PVince81) [inline lib/private/DB/PostgreSqlMigrator.php]: did you try doing this earlier, maybe within the schema diff ? or is the word "BIGSERIAL" generated very late ?

File: lib/private/DB/PostgreSqlMigrator.php [hunk 3]
```diff
@@ -52,4 +52,32 @@ protected function getDiff(Schema $targetSchema, \Doctrine\DBAL\Connection $conn
...
+				$alterTable = $matches[1];
+				$sql = $alterTable . 'BIGINT';
```
Comment (PVince81) [inline lib/private/DB/PostgreSqlMigrator.php]: strange, you just append it to the end ? wouldn't BIGSERIAL appear in the middle of a statement sometimes ?

File: apps/dav/appinfo/Migrations/Version20170711193427.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+		if ($schema->hasTable("${prefix}properties")) {
+			$table = $schema->getTable("{$prefix}properties");
+
+			$idColumn = $table->getColumn('id');
+			if ($idColumn){
+				$idColumn->setType(Type::getType(Type::BIGINT));
+				
+				// Fixup postgres autoincrement
+				if ($this->connection->getDatabasePlatform() instanceof PostgreSqlPlatform){
+					$default = sprintf("nextval('%s_%s_seq'::regclass)", $table->getName(), $idColumn->getName());
```
Comment (PVince81) [inline apps/dav/appinfo/Migrations/Version20170711193427.php]: sad if we can't do this inside of `PostgreSqlPlatform`. Will we need this statement in more places ? Worst case, put it as a utility function in `PostgreSqlPlatform` then for reusability ?

**Comment by PVince81** [comment]
Ok, I just saw your comment that you already looked into some way to do it in Doctrine. Since it's not possible, let's see if we can make parts of this code reusable, especially the logic for `setDefault`. I expect we'll need it in subsequent PRs when addressing more ids.

File: lib/private/DB/PostgreSqlMigrator.php (hunk 3 above)
Comment (VicDeo) [inline lib/private/DB/PostgreSqlMigrator.php]: @PVince81 Pure regexp magic. I capture everything into `$matches`
 `(something)(anotherthing)`
when the input string is `somethinganotherthing`
`$matches[0]` will be `somethinganotherthing`
`$matches[1]` will be `something`
`$matches[2]` will be `anotherthing`

in this case 
`$matches[1]` is `ALTER TABLE whatever ALTER somecolumn TYPE `
and
`$matches[2]` is `BIGSERIAL`

Comment (PVince81) [inline lib/private/DB/PostgreSqlMigrator.php]: what if there is more than one column in a single ALTER statement, or does Doctrine always split these ?

File: lib/private/DB/PostgreSqlMigrator.php (hunk 2 above)
Comment (VicDeo) [inline lib/private/DB/PostgreSqlMigrator.php]: @PVince81 sqls are generated in line 68  `$sqls = $schemaDiff->toSql($connection->getDatabasePlatform());`

File: lib/private/DB/PostgreSqlMigrator.php (hunk 3 above)
Comment (VicDeo) [inline lib/private/DB/PostgreSqlMigrator.php]: @PVince81 schemaDiff generates statements per column https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Platforms/PostgreSqlPlatform.php#L521

File: tests/lib/DB/OCPostgreSqlPlatformTest.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+			true
+		);
+		
+		$this->assertNotContains(
+			'ALTER TABLE poor_yorick ALTER id DROP DEFAULT',
+			$sqlStatements,
+			true
+		);
+		
+		var_dump($sqlStatements);
```
Comment (VicDeo) [inline tests/lib/DB/OCPostgreSqlPlatformTest.php]: Leftover

**Comment by VicDeo** [comment]
@PVince81 
Next level. I override `PostgreSqlPlatform::getAlterTableSQL` from Doctrine and cleanup it's output.
A basic unit test to validate a result comes as a bonus.

**Comment by VicDeo** [comment]
@PVince81 
> let's see if we can make parts of this code reusable, especially the logic for `setDefault`

done.

**Comment by PVince81** [comment]
I'm fine with this but would like to know whether @DeepDiver1975 is ok with the various DB gymnastics

File: apps/files_sharing/appinfo/database.xml (hunk 1 above)
Comment (DeepDiver1975) [inline apps/files_sharing/appinfo/database.xml]: I'm thinking about - for the sake of clearness - to kill all database.xml files and add them as migrations ....

**Review by DeepDiver1975 — DISMISSED** [review]

**Comment by DeepDiver1975** [comment]
Rebased - this was really out dated - please always update your branches - thx

**Comment by DeepDiver1975** [comment]
bildschirmfoto von 2017-07-31 09-11-00

**Comment by DeepDiver1975** [comment]
if stuff above is addressed: 👍

**Comment by PVince81** [comment]
Talked with @DeepDiver1975: let's not backport the bigint fileid thing. People should upgrade, especially considering that we're preparing a way already to upgrade from 8.2 to 10 (upgrade from 9.0 to 10 is already possible):

**Comment by VicDeo** [comment]
@DeepDiver1975 updated
@PVince81 so, this change is not for OC 10 or we will release 10 from master?

**Comment by PVince81** [comment]
@VicDeo sorry for the confusion. This is to be released for 10.0.3, so backport to stable10. But nothing for OC <= 9.1

**Comment by VicDeo** [comment]
Rebased for CI

**Comment by VicDeo** [comment]
@PVince81 @DeepDiver1975 Tests are passed. Can't merge without approval

**Comment by PVince81** [comment]
@VicDeo please backport to stable10

**Comment by VicDeo** [comment]
Stable10: https://github.com/owncloud/core/issues/28581

**Pull request merged**